cnd/src/btsieve/mod.rs

```diff
@@ -88,39 +88,37 @@ where
 /// This function yields all blocks as part of its process.
 /// This function returns the block-hashes of all visited blocks.
 async fn walk_back_until<C, P, B, H>(
-    stop_condition: P,
+    should_stop_here: P,
+    starting_block: B,
     connector: &C,
     co: &Co<B>,
-    block: B,
 ) -> anyhow::Result<HashSet<H>>
 where
     C: BlockByHash<Block = B, BlockHash = H>,
     P: Fn(&B) -> bool,
-    B: BlockHash<H> + PreviousBlockHash<H> + Clone,
+    B: BlockHash<H> + PreviousBlockHash<H>,
     H: Eq + Hash + Copy,
 {
-    let mut seen_blocks: HashSet<H> = HashSet::new();
-    let mut blockhash = block.block_hash();
+    let mut seen_blocks = HashSet::new();
 
-    co.yield_(block.clone()).await;
-    seen_blocks.insert(blockhash);
-
-    if stop_condition(&block) {
-        return Ok(seen_blocks);
-    } else {
-        blockhash = block.previous_block_hash();
-    }
+    let mut current_blockhash = starting_block.block_hash();
+    let mut current_block = starting_block;
 
     loop {
-        let block = connector.block_by_hash(blockhash).await?;
-        co.yield_(block.clone()).await;
-        seen_blocks.insert(blockhash);
+        seen_blocks.insert(current_blockhash);
 
-        if stop_condition(&block) {
+        // compute variables before consuming the block by yielding it
```

Comment:
```suggestion
        // yielding the block consumes it
```

Reply:
Use comments that say _why_ we do something not _what_ we are doing :)

Reply:
Well, the comment says _why_ we are doing this here and not further down. I am not sure how reducing the comment to your suggestion makes that more obvious as it removes the link between the comment and where it is placed.

If I would read: "yielding the block consumes it", my first thought would be:

If the comment concerns the yielding, why is it not on top of the yield statement?

And the second one would be: So what? Moving is the default in Rust.

The reason I added the comment is because it is not immediately obvious that we have to compute these values before we yield even though we only need them afterwards.